text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,362 @@
+- Feature Name: Threadsafe Handles
+- Start Date: 2020-07-21
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Two new `Send` and `Sync` features are introduced: `EventQueue: Send + Sync` and `Persistent: Send`. These features work together to allow multi-threaded modules to interact with the JavaScript main thread.
+
+* `EventQueue`: Threadsafe handle that allows sending closures to be executed on the main thread.
+* `Persistent<T>`: Opaque handle that can be sent across threads, with inner contents only accessible on the main thread.
+
+# Motivation
+[motivation]: #motivation
+
+High level event handlers have gone through [several](https://github.com/neon-bindings/rfcs/pull/25) [iterations](https://github.com/neon-bindings/rfcs/pull/28) without quite providing [ideal](https://github.com/neon-bindings/rfcs/issues/31) ergonomics or [safety](https://github.com/neon-bindings/neon/issues/551).
+
+The Threadsafe Handlers feature attempts to decompose the high level `EventHandler` API into lower level, more flexible primitives.
+
+These primitives can be used to build a higher level and _safe_ `EventHandler` API allowing experimentation with patterns outside of `neon`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Neon provides a smart pointer type `Handle<'a, T: Value>` for referencing values on the JavaScript heap. These references are bound by the lifetime of `Context<'a>` to ensure they can only be used while the VM is locked in a synchronous neon function. These are neither `Send` or `Sync`.
+
+### `neon::handle::Persistent<T>`
+
+As a developer, I may want to retain a reference to a JavaScript value while returning control back to the VM.
+
+`Handle<'_, _>` may not outlive the context that created them.
+
+```rust
+// Does not compile because `cb` cannot be sent across threads
+fn thread_log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?;
+
+    std::thread::spawn(move || {
+        println!("{:?}", cb);
+    });
+
+    cx.undefined()
+}
+```
+
+A persistent reference may be created from a `Handle<'_, _>` which can be sent across threads.
+
+```rust
+// Compiles
+fn thread_log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+
+    std::thread::spawn(move || {
+        // Panics when `cb` is dropped.
```

Comment:
Not critical, but putting this example here in this order is confusing for the reader of the RFC, since you haven't yet explained that persistent handles have to be returned to the JS thread to be properly dropped.